File: server/internal/trust/trust.go
package trust

import "math"

// GraphView is the viewer's entitled 2-hop slice (docs/protocol.md §6,
// GET /graph/2hop): their own follows, the follow list of each account they
// follow, plus their private mutes. This mirrors packages/core/src/trust.ts;
// the shared trust-graph-01 vector keeps the two implementations honest.
type GraphView struct {
	Follows map[string][]string
	Mutes   []string
}

// Params are the tunable constants of the trust computation
// (docs/trust-and-reach.md §6). Hop cap 2 is structural, not a parameter.
type Params struct {
	PerHopDecay     float64
	MultiPathSumCap float64
}

// DefaultParams are the published reference defaults.
var DefaultParams = Params{PerHopDecay: PerHopDecay, MultiPathSumCap: MultiPathSumCap}

func toSet(items []string) map[string]bool {
	set := make(map[string]bool, len(items))
	for _, it := range items {
		set[it] = true
	}
	return set
}

// SubjectiveTrust computes subjective_trust(viewer, author) per
// docs/trust-and-reach.md §1: direct follow = 1.0; each two-hop path adds
// PerHopDecay; paths sum, capped at MultiPathSumCap. Mutes are a hard zero
// that also prunes propagation through the muted account. Follow lists are
// deduped so a repeated edge can never double-count.
//
// Self-trust is not defined by the spec — callers always show the viewer
// their own records; SubjectiveTrust panics on viewer == author exactly as
// the TS reference throws.
func SubjectiveTrust(viewer, author string, graph GraphView, p Params) float64 {
	if author == viewer {
		panic("self-trust is not defined; treat own content as always visible")
	}
	muted := toSet(graph.Mutes)
	if muted[author] {
		return 0
	}
	weight := 0.0
	for mid := range toSet(graph.Follows[viewer]) {
		if muted[mid] { // hard zero prunes the muted account's outbound edges
			continue
		}
		if mid == author {
			weight += 1.0
			continue
		}
		if toSet(graph.Follows[mid])[author] {
			weight += p.PerHopDecay
		}
	}
	return math.Min(weight, p.MultiPathSumCap)
}

// TrustMap computes trust for every account reachable within the hop cap in
// one pass. The viewer is never a key; muted accounts are absent (their
// trust is the implicit zero).
func TrustMap(viewer string, graph GraphView, p Params) map[string]float64 {
	muted := toSet(graph.Mutes)
	out := make(map[string]float64)
	for mid := range toSet(graph.Follows[viewer]) {
		if muted[mid] {
			continue
		}
		out[mid] += 1.0
		for far := range toSet(graph.Follows[mid]) {
			if far == viewer || muted[far] {
				continue
			}
			out[far] += p.PerHopDecay
		}
	}
	for k, v := range out {
		out[k] = math.Min(v, p.MultiPathSumCap)
	}
	return out
}

// DailyBudget is the cold-outreach budget formula of docs/trust-and-reach.md
// §3: (base + k×ln(1+Σ inbound_trust)) × standing. Mirrors
// packages/core/src/budgets.ts; asserted by the budgets-01 vector. Negative
// inbound trust is clamped to zero. (Used by the Phase 4 budget mechanics;
// the math lands here with its vector.)
func DailyBudget(base, inboundTrust, k, standing float64) float64 {
	return (base + k*math.Log(1+math.Max(0, inboundTrust))) * standing
}

File: packages/core/src/trust.ts
/**
 * Subjective trust computation (docs/trust-and-reach.md §§1–2). This is the
 * published math every client re-runs before rendering anything as trusted;
 * the Go server mirrors it for candidate ranking, and simlab exercises this
 * exact implementation. Shared vectors (trust-graph-01) keep all three honest.
 *
 * The viewer's own content is outside trust: trust(viewer, viewer) is not
 * defined by the spec — callers always show the viewer their own records.
 */
import { CONSTANTS } from "./constants.js";

/**
 * The viewer's entitled 2-hop slice: their own follows, plus the follow list
 * of each account they follow (GET /graph/2hop), plus their private mutes.
 */
export interface GraphView {
  follows: Record<string, readonly string[]>;
  mutes?: readonly string[];
}

export interface TrustConstants {
  per_hop_decay: number;
  multi_path_sum_cap: number;
  feed_surface_threshold: number;
}

/**
 * subjective_trust(viewer, author): direct follow = 1.0; each two-hop path
 * = decay; paths sum, capped at multi_path_sum_cap. Mutes are a hard zero
 * that also prunes propagation through the muted account. Hop cap 2 is
 * structural: anything further has no path.
 */
export function subjectiveTrust(
  viewer: string,
  author: string,
  graph: GraphView,
  constants: TrustConstants = CONSTANTS,
): number {
  if (author === viewer) throw new Error("self-trust is not defined; treat own content as always visible");
  const muted = new Set(graph.mutes ?? []);
  if (muted.has(author)) return 0;
  // Follow lists are sets by construction; dedupe defensively so a repeated
  // edge can never double-count into trust.
  const viewerFollows = new Set(graph.follows[viewer] ?? []);
  let weight = 0;
  for (const mid of viewerFollows) {
    if (muted.has(mid)) continue; // hard zero prunes the muted account's outbound edges
    if (mid === author) {
      weight += 1.0;
      continue;
    }
    if (new Set(graph.follows[mid] ?? []).has(author)) weight += constants.per_hop_decay;
  }
  return Math.min(weight, constants.multi_path_sum_cap);
}

/** Trust for every account reachable within the hop cap, from one pass. */
export function trustMap(
  viewer: string,
  graph: GraphView,
  constants: TrustConstants = CONSTANTS,
): Record<string, number> {
  const muted = new Set(graph.mutes ?? []);
  const out: Record<string, number> = {};
  for (const mid of new Set(graph.follows[viewer] ?? [])) {
    if (muted.has(mid)) continue;
    out[mid] = (out[mid] ?? 0) + 1.0;
    for (const far of new Set(graph.follows[mid] ?? [])) {
      if (far === viewer || muted.has(far)) continue;
      out[far] = (out[far] ?? 0) + constants.per_hop_decay;
    }
  }
  for (const k of Object.keys(out)) {
    out[k] = Math.min(out[k]!, constants.multi_path_sum_cap);
  }
  return out;
}

/** effective_trust = subjective_trust × standing (standing ∈ [0,1], default 1). */
export function effectiveTrust(subjective: number, standing = 1.0): number {
  if (standing < 0 || standing > 1) throw new Error("standing out of range [0,1]");
  return subjective * standing;
}

/**
 * Feed buckets (trust-and-reach §2): ≥ threshold ranks normally; a positive
 * sub-threshold score exists but doesn't surface unprompted; no path at all
 * is pull-only. `directFollow` (protocol §9.3, trust-and-reach §5 invariant
 * 3): an author the viewer directly follows ranks normally regardless of
 * effective trust — a standing penalty never severs a chosen edge.
 */
export type FeedBucket = "normal" | "below-threshold" | "no-path";

export function feedBucket(
  effective: number,
  constants: TrustConstants = CONSTANTS,
  directFollow = false,
): FeedBucket {
  if (directFollow) return "normal";
  if (effective >= constants.feed_surface_threshold) return "normal";
  if (effective > 0) return "below-threshold";
  return "no-path";
}
